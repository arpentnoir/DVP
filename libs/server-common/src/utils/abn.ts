const weights = [10, 1, 3, 5, 7, 9, 11, 13, 15, 17, 19];

/**
 *
 * @param rawAbn Validates the ABN format based on https://abr.business.gov.au/Help/AbnFormat
 *
 * The 11 digit ABN is structured as a 9 digit identifier with two leading check digits. The leading check digits are derived using a modulus 89 (remainder after dividing by 89) calculation.
 *
 * To verify an ABN:
 *  1. Subtract 1 from the first (left-most) digit of the ABN to give a new 11 digit number
 *  2. Multiply each of the digits in this new number by a "weighting factor" based on its position as shown in the table below
 *  3. Sum the resulting 11 products
 *  4. Divide the sum total by 89, noting the remainder
 *  5. If the remainder is zero the number is a valid ABN
 *
 * For example, to check if 51 824 753 556 is a valid ABN:
 *
 *  1. Subtract 1 from the first (left-most) digit (5) to give 41 824 753 556
 *  2. Multiply each of the digits in 41 824 753 556 by the "weighting factor" based on its position as shown in the table below
 *  3. Sum (Digit * weight) to give a total of 534
 *  4. Divide 534 by 89 giving 6 with zero remainder.
 *  5. As the remainder is zero, 51 824 753 556 is a valid ABN.

 * @returns boolean
 */
export function isValidABN(rawAbn: string | number): boolean {
  if (!rawAbn) {
    return false;
  }

  // strip non-alphanumeric characters
  const abn = rawAbn.toString().replace(/[^a-z\d]/gi, '');

  // check if length is 11 digits
  if (abn.length !== 11) {
    return false;
  }

  // apply ato check method
  let sum = 0;
  weights.forEach((weight, position) => {
    const digit = Number(abn[position]) - (position ? 0 : 1);
    sum += weight * digit;
  });

  const checksum = sum % 89;

  return checksum === 0;
}
