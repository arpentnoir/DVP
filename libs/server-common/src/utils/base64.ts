export const encode = (data: string) => {
  const buff = Buffer.from(data);
  return buff.toString('base64');
};

export const decode = (base64String: string) => {
  const buff = Buffer.from(base64String, 'base64');
  return buff.toString('ascii');
};
