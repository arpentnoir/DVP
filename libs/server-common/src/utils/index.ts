export * from './base64';
export * from './date';
export * from './environment';
export * as kms from './kms';
export * from './logger';
export * from './s3-adapter';
export * from './test-helpers';
export * from './uuid';
export * from './vc';
