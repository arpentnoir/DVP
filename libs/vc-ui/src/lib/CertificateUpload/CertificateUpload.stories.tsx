import type { ComponentStory, ComponentMeta } from '@storybook/react';
import { CertificateUpload } from './CertificateUpload';

const Story: ComponentMeta<typeof CertificateUpload> = {
  component: CertificateUpload,
  title: 'CertificateUpload',
};
export default Story;

const Template: ComponentStory<typeof CertificateUpload> = (args) => (
  <CertificateUpload {...args} />
);

export const Primary = Template.bind({});
