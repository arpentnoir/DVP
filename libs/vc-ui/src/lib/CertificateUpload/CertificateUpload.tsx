import CloudUploadIcon from '@mui/icons-material/CloudUpload';
import {
  Alert,
  Box,
  Input,
  InputLabel,
  Stack,
  SxProps,
  useMediaQuery,
  useTheme,
} from '@mui/material';
import React, {
  DragEvent,
  FormEvent,
  useEffect,
  useRef,
  useState,
} from 'react';
import { Button } from '../Button';
import { Text } from '../Text';
interface ICertificateUpload {
  handleFiles: (files: FileList) => void;
  errorMessage?: string;
  acceptMessage?: string;
  style?: { mobile?: SxProps; desktop?: SxProps };
}

export const CertificateUpload = ({
  handleFiles,
  errorMessage,
  acceptMessage,
  style,
}: ICertificateUpload) => {
  const [dragActive, setDragActive] = useState(false);
  const inputRef = useRef<HTMLInputElement>(null);
  const errorRef = useRef<HTMLParagraphElement>(null);

  const theme = useTheme();

  useEffect(() => {
    if (errorMessage && errorRef.current) {
      errorRef.current.focus();
    }
  }, [errorMessage]);

  const handleDrag = function (
    event: DragEvent<HTMLDivElement | HTMLFormElement>
  ) {
    event.preventDefault();
    event.stopPropagation();
    if (event.type === 'dragenter' || event.type === 'dragover') {
      setDragActive(true);
    } else if (event.type === 'dragleave') {
      setDragActive(false);
    }
  };

  const handleDrop = function (event: DragEvent<HTMLDivElement>) {
    event.preventDefault();
    event.stopPropagation();
    setDragActive(false);
    if (event?.dataTransfer?.files && event.dataTransfer.files[0]) {
      handleFiles(event.dataTransfer.files);
    }
  };

  const handleChange = function (
    event: FormEvent<HTMLInputElement | HTMLTextAreaElement>
  ) {
    event.preventDefault();
    const target = event?.target as HTMLInputElement;
    if (target.files && target.files[0]) {
      handleFiles(target.files);
    }
  };

  const onKeyUp = (event: any) => {
    event.preventDefault();
    if (event.key === ' ' || event.key === 'Enter') {
      inputRef?.current?.click();
    }
  };

  const onButtonClick = (event: any) => {
    event.preventDefault();
    inputRef?.current?.click();
  };

  const textColor = 'black';

  const matches = useMediaQuery(`(min-width:${theme.breakpoints.values.sm}px)`);

  return (
    <React.Fragment>
      <Box
        onDragEnter={handleDrag}
        id="certificate-upload"
        sx={{
          textAlign: 'center',
        }}
      >
        {!matches && (
          <Stack
            sx={{
              display: { xs: 'flex' },
              justifyContent: 'center',
              // Override style
              ...style?.mobile,
            }}
          >
            <Stack width="100%" gap="56px" alignItems="center">
              {errorMessage && (
                <Alert
                  severity="error"
                  id="vc-upload-error"
                  ref={errorRef}
                  aria-label={errorMessage}
                  tabIndex={0}
                  sx={{
                    backgroundColor: 'rgba(173, 26, 31, 0.04)',
                    width: '100%',
                    height: '60px',
                    justifyContent: 'center',
                    alignItems: 'center',
                    border: '1px solid #AD1A1F',
                    borderRadius: '8px',
                    color: '#AD1A1F',
                    whiteSpace: 'normal',
                  }}
                >
                  {errorMessage}
                </Alert>
              )}
              <Stack spacing={2} alignItems="center" width="100%">
                {/* TODO: A decision needs to be made whether we are providing the functionality to scan a QRCode */}
                <Button
                  sx={{
                    maxWidth: '510px',
                    width: '100%',
                    height: '42px',
                    backgroundColor: '#072243',
                    ':hover': {
                      boxShadow: 'none',
                      backgroundColor: '#092e5a',
                    },
                  }}
                  variant="text"
                  label={'Scan QR code'}
                  textProps={{ variant: 'h4', color: 'white', tabIndex: -1 }}
                ></Button>
                <Box sx={{ color: 'black' }}>
                  <Text
                    sx={{
                      display: 'inline',
                    }}
                    variant="body1"
                    tabIndex={-1}
                  >
                    Or &nbsp;
                  </Text>
                  <Box
                    role="button"
                    aria-controls="filename"
                    tabIndex={0}
                    aria-label={`Select a file. ${acceptMessage}`}
                    onKeyUp={onKeyUp}
                    onClick={onButtonClick}
                    sx={{
                      display: 'inline',
                      borderBottom: '1px solid',
                      ':hover': { cursor: 'pointer' },
                    }}
                  >
                    Select a file
                  </Box>
                </Box>
              </Stack>
            </Stack>
          </Stack>
        )}
        <Input
          inputRef={inputRef}
          type="file"
          id="input-file-upload"
          inputProps={{ 'data-testid': 'input-file-upload' }}
          onChange={handleChange}
          sx={{ display: 'none' }}
        />
        <InputLabel
          id="label-file-upload"
          htmlFor="input-file-upload"
          aria-label="Upload file"
          sx={{ height: '100%', display: { xs: 'hidden', sm: 'block' } }}
        >
          {matches && (
            <Stack
              sx={{
                height: '100%',
                alignItems: 'center',
                justifyContent: 'space-evenly',
                borderWidth: '2px',
                borderRadius: '8px',
                borderStyle: 'dashed',
                borderColor: errorMessage ? '#AD1A1F' : '#cbd5e1',
                backgroundColor: errorMessage
                  ? 'rgba(173, 26, 31, 0.04)'
                  : 'white',
                // Override style
                ...style?.desktop,
              }}
            >
              <Stack
                sx={{
                  display: 'flex',
                  flexDirection: 'column',
                  alignItems: 'center',
                  padding: '20px 100px',
                  gap: '14px',
                  color: textColor,
                }}
              >
                <CloudUploadIcon
                  fontSize="large"
                  sx={{
                    fill: '#0C1B2E',
                  }}
                />
                <Text
                  fontWeight="bold"
                  color={textColor}
                  variant="h6"
                  tabIndex={-1}
                >
                  Drop your document to verify
                </Text>
                <Box>
                  <Text
                    sx={{
                      display: 'inline',
                    }}
                    variant="body1"
                    tabIndex={-1}
                  >
                    Or &nbsp;
                  </Text>
                  <Box
                    role="button"
                    aria-controls="filename"
                    tabIndex={0}
                    aria-label={`Select a file. ${acceptMessage}`}
                    data-testid="file-select-button"
                    onKeyUp={onKeyUp}
                    sx={{
                      display: 'inline',
                      borderBottom: '1px solid',
                      ':hover': { cursor: 'pointer' },
                    }}
                  >
                    Select a file
                  </Box>
                </Box>

                {acceptMessage && (
                  <Text variant="body1" tabIndex={-1}>
                    {acceptMessage}
                  </Text>
                )}
                {errorMessage && (
                  <Alert
                    severity="error"
                    id="vc-upload-error"
                    ref={errorRef}
                    aria-label={errorMessage}
                    data-testid="vc-upload-error"
                    tabIndex={1}
                    sx={{
                      backgroundColor: 'transparent',
                      justifyContent: 'center',
                      alignItems: 'center',
                      border: 'none',
                      color: '#AD1A1F',
                      whiteSpace: 'normal',
                    }}
                  >
                    {errorMessage}
                  </Alert>
                )}
              </Stack>
            </Stack>
          )}
        </InputLabel>
        {dragActive && (
          <Box
            sx={{
              position: 'absolute',
              height: ' 100%',
              borderRadius: '1rem',
              top: '0px',
              right: '0px',
              bottom: '0px',
              left: '0px',
            }}
            id="drag-file-element"
            onDragEnter={handleDrag}
            onDragExit={handleDrag}
            onDragOver={handleDrag}
            onDrop={handleDrop}
          />
        )}
      </Box>
    </React.Fragment>
  );
};
