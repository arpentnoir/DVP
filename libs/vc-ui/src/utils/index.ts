export * from './clipboard';
export * from './fileToBase64';
export * from './RouterWrapper';
export * from './vc';
