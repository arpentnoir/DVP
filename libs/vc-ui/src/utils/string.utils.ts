export const capitaliseFirstLetter = (text: string) => {
  if (text) {
    return text.charAt(0).toUpperCase() + text.slice(1);
  }
  return text;
};

export const constructPath = (path: string[]) => {
  return path.reduce((previousValue, path) => {
    return path ? `${previousValue}/${path}` : '';
  }, '');
};
