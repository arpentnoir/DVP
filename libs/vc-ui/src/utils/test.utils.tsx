import { useTheme } from '@mui/material';
import { render as TLRender } from '@testing-library/react';
import mediaQuery from 'css-mediaquery';
import { createMemoryHistory } from 'history';
import { Location, Navigator, Router } from 'react-router-dom';

export const render = (
  element: React.ReactNode,
  options?: {
    location?: Partial<Location> | string;
    navigator?: Navigator;
  }
) => {
  const defaultOptions = {
    location: '',
    navigator: createMemoryHistory(),
  };

  return TLRender(
    <Router {...defaultOptions} {...options}>
      {element}
    </Router>
  );
};

export const setScreenSize =
  (breakpoint: 'xs' | 'sm' | 'md' | 'lg' | 'xl') => (query: string) => {
    const theme = useTheme();
    return {
      matches: mediaQuery.match(query, {
        width: theme.breakpoints.values[breakpoint],
      }),
      addListener: () => {
        return;
      },
      removeListener: () => {
        return;
      },
    };
  };
