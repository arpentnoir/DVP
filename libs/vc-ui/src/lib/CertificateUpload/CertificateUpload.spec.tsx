import { fireEvent, render, waitFor } from '@testing-library/react';
import { CertificateUpload } from './CertificateUpload';
import { setScreenSize } from '../../utils/test.utils';

const handleFiles = jest.fn();
const ACCEPT_MESSAGE = 'Max 3MB';
const ERROR_MESSAGE = 'Something went wrong';

describe('CertificateUpload', () => {
  describe('Desktop', () => {
    beforeAll(() => {
      (window.matchMedia as any) = setScreenSize('sm');
    });

    it('should render correctly', () => {
      const { baseElement } = render(
        <CertificateUpload
          handleFiles={handleFiles}
          acceptMessage={ACCEPT_MESSAGE}
          errorMessage=""
        />
      );

      expect(baseElement).toMatchSnapshot('Desktop');
    });

    it('should display error message if present', () => {
      let baseElement = render(
        <CertificateUpload
          handleFiles={handleFiles}
          acceptMessage={ACCEPT_MESSAGE}
          errorMessage=""
        />
      );

      expect(baseElement.queryByText('Something went wrong')).toBeFalsy();

      baseElement = render(
        <CertificateUpload
          handleFiles={handleFiles}
          acceptMessage={ACCEPT_MESSAGE}
          errorMessage={ERROR_MESSAGE}
        />
      );

      expect(baseElement.queryByText('Something went wrong')).toBeTruthy();
    });
  });

  describe('Mobile', () => {
    beforeAll(() => {
      (window.matchMedia as any) = setScreenSize('xs');
    });

    it('should render correctly', () => {
      const { baseElement } = render(
        <CertificateUpload
          handleFiles={handleFiles}
          acceptMessage={ACCEPT_MESSAGE}
          errorMessage=""
        />
      );

      expect(baseElement).toMatchSnapshot('Mobile');
    });

    it('should display error message if present', () => {
      let baseElement = render(
        <CertificateUpload
          handleFiles={handleFiles}
          acceptMessage={ACCEPT_MESSAGE}
          errorMessage=""
        />
      );

      expect(baseElement.queryByText('Something went wrong')).toBeFalsy();

      baseElement = render(
        <CertificateUpload
          handleFiles={handleFiles}
          acceptMessage={ACCEPT_MESSAGE}
          errorMessage={ERROR_MESSAGE}
        />
      );

      expect(baseElement.queryByText('Something went wrong')).toBeTruthy();
    });
  });

  it('should call handleFiles on upload', async () => {
    const baseElement = render(
      <CertificateUpload
        handleFiles={handleFiles}
        acceptMessage={ACCEPT_MESSAGE}
        errorMessage=""
      />
    );

    const uploadElement = baseElement.getByTestId('input-file-upload');

    await waitFor(() =>
      fireEvent.change(uploadElement, {
        target: { files: ['{"test": "hello"}'] },
      })
    );

    expect(handleFiles).toHaveBeenCalledWith(['{"test": "hello"}']);
  });
});
