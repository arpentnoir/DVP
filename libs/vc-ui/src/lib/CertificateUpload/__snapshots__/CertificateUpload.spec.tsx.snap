// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CertificateUpload Desktop should render correctly: Desktop 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-xi606m"
      id="certificate-upload"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-1g6u8k6-MuiInputBase-root-MuiInput-root"
      >
        <input
          class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          data-testid="input-file-upload"
          id="input-file-upload"
          type="file"
          value=""
        />
      </div>
      <label
        aria-label="Upload file"
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-1l91s8u-MuiFormLabel-root-MuiInputLabel-root"
        for="input-file-upload"
        id="label-file-upload"
      >
        <div
          class="css-1kz0kwo-MuiStack-root"
        >
          <div
            class="css-1fbeulj-MuiStack-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-13frft-MuiSvgIcon-root"
              data-testid="CloudUploadIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM14 13v4h-4v-4H7l5-5 5 5h-3z"
              />
            </svg>
            <h6
              class="MuiTypography-root MuiTypography-h6 css-n8wto6-MuiTypography-root"
              tabindex="-1"
            >
              Drag and drop your document to Verify
            </h6>
            <div
              class="MuiBox-root css-0"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-1sx1eu0-MuiTypography-root"
                tabindex="-1"
              >
                Or  
              </p>
              <div
                aria-controls="filename"
                aria-label="Select a file. Max 3MB"
                class="MuiBox-root css-13imy7l"
                data-testid="file-select-button"
                role="button"
                tabindex="0"
              >
                Upload a Document
              </div>
            </div>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              tabindex="-1"
            >
              Max 3MB
            </p>
          </div>
        </div>
      </label>
    </div>
  </div>
</body>
`;

exports[`CertificateUpload Mobile should render correctly: Mobile 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-xi606m"
      id="certificate-upload"
    >
      <div
        class="css-1g3jqqf-MuiStack-root"
      >
        <div
          class="css-1wrk7t-MuiStack-root"
        >
          <div
            class="css-1i78psr-MuiStack-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1eoulhv-MuiButtonBase-root-MuiButton-root"
              data-testid="button:Scan QR code"
              tabindex="0"
              type="button"
            >
              <div
                class="css-doi3ij-MuiStack-root"
              >
                <h4
                  class="MuiTypography-root MuiTypography-h4 css-5ukp3i-MuiTypography-root"
                  data-testid="button:Scan QR code:label"
                  tabindex="-1"
                >
                  Scan QR code
                </h4>
              </div>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <div
              class="MuiBox-root css-1xu3tth"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-1sx1eu0-MuiTypography-root"
                tabindex="-1"
              >
                Or  
              </p>
              <div
                aria-controls="filename"
                aria-label="Select a file. Max 3MB"
                class="MuiBox-root css-13imy7l"
                role="button"
                tabindex="0"
              >
                Select a file
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-1g6u8k6-MuiInputBase-root-MuiInput-root"
      >
        <input
          class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          data-testid="input-file-upload"
          id="input-file-upload"
          type="file"
          value=""
        />
      </div>
      <label
        aria-label="Upload file"
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-1l91s8u-MuiFormLabel-root-MuiInputLabel-root"
        for="input-file-upload"
        id="label-file-upload"
      />
    </div>
  </div>
</body>
`;
